app/src/crypto.c

```diff
@@ -584,3 +612,288 @@ zxerr_t crypto_fillMASP(uint8_t *buffer, uint16_t bufferLen, uint16_t *cmdRespon
+    MEMCPY(signature + HASH_LEN, sbar, HASH_LEN);
+
+    return zxerr_ok;
+}
+
+zxerr_t crypto_sign_spends_sapling(const parser_tx_t *txObj, keys_t *keys, uint8_t *output, uint16_t outputLen, uint16_t *responseLen) {
+    zemu_log_stack("crypto_signspends_sapling");
+    if (txObj->transaction.sections.maspTx.data.sapling_bundle.n_shielded_spends == 0) {
+        return zxerr_ok;
+    }
+
+    MEMZERO(output, outputLen);
+    // Get Signature hash
+    uint8_t sign_hash[HASH_LEN] = {0};
+    signature_hash(txObj, sign_hash);
+
+    uint8_t signature[2 * HASH_LEN] = {0};
+    uint8_t alpha[KEY_LENGTH] = {0};
+    const uint8_t *spend = txObj->transaction.sections.maspBuilder.builder.sapling_builder.spends.ptr;
+    uint16_t spendLen = 0;
+
+    for (uint64_t i = 0; i < txObj->transaction.sections.maspBuilder.builder.sapling_builder.n_spends; i++) {
+        // Get spend description and alpha
+        spend += spendLen;
+        MEMCPY(alpha, spend + ALPHA_OFFSET, KEY_LENGTH);
+
+        CHECK_ZXERR(sign_sapling_spend(keys, alpha, sign_hash, signature));
+
+        // Copy signature to output
+        MEMCPY(output + i * MASP_SIG_LEN, signature, MASP_SIG_LEN);
+
+        // Get this spend lenght to get next one
+        getSpendDescriptionLen(spend, &spendLen);
+
+        *responseLen += MASP_SIG_LEN;
+    }
+
+    return zxerr_ok;
+}
+
+
+zxerr_t checkSpends(const parser_tx_t *txObj, keys_t *keys) {
+    uint8_t *builder_spend = (uint8_t *)txObj->transaction.sections.maspBuilder.builder.sapling_builder.spends.ptr;
+    uint8_t *tx_spend = (uint8_t *)txObj->transaction.sections.maspTx.data.sapling_bundle.shielded_spends.ptr;
+    uint16_t spendLen = 0;
+
+    for (uint32_t i = 0; i < txObj->transaction.sections.maspBuilder.builder.sapling_builder.n_spends; i++) {
+        builder_spend += spendLen;
+        tx_spend += SHIELDED_SPENDS_LEN;
+
+        //check rk
+        uint8_t alpha[KEY_LENGTH] = {0};
+        MEMCPY(alpha, builder_spend + ALPHA_OFFSET, KEY_LENGTH);
+        uint8_t rk[KEY_LENGTH] = {0};
+        CHECK_PARSER_OK(computeRk(keys, alpha, rk));
+        if(MEMCMP(rk, tx_spend + CV_LEN + NULLIFIER_LEN, RK_LEN) != 0) {
+            return zxerr_invalid_crypto_settings;
+        }
+
+        //check cv (computation validaded in cpp_tests
+        uint8_t cv[KEY_LENGTH] = {0};
+        uint8_t *note = builder_spend + EXTENDED_FVK_LEN + DIVERSIFIER_LEN;
+        uint8_t *rcv = (uint8_t*)txObj->transaction.sections.maspBuilder.metadata.spend_rcvs.ptr + (i * KEY_LENGTH);
+        uint64_t value = 0;
+        MEMCPY(&value, note + IDENTIFIER_LEN, sizeof(uint64_t));
+        CHECK_PARSER_OK(computeValueCommitment(value, rcv, note, cv));
+        if(MEMCMP(cv, tx_spend, CV_LEN) != 0) {
+            return zxerr_invalid_crypto_settings;
+        }
+    
+        CHECK_PARSER_OK(getSpendDescriptionLen(builder_spend, &spendLen));
+    }
+    return zxerr_ok;
+}
+
+zxerr_t checkOutputs(const parser_tx_t *txObj) {
```

Comment:
Similar to `checkSpends` comments. Before reading from pointers `builder_output` and `tx_output`, please check that you still have bytes